<pre><b>
Protocol - set of rules to communicate between machines.
Example - Grammar is a protocol for a language. You have to follow grammar while speaking so that others can understand your communication.
USB port - follows usb protocol so that when we plug usb, it reads from usb.

Simple Mail Transfer Protocol - to communicate with a mail server, we have to follow SMTP. This handles sending and trasferring mail.

SMTP is a part of python network library. It is a part of Application Layer.

SSL/TLS - standard encryption presents in TCP(Transmission Control Protocol) communication.

SSL - Secure Socket Layer
TLS - Transport Layer Security

Steps to communicate using SMTP:
    1. Create SMTP client
    2. We have to have host(domain name), port number
</pre></b>

In [1]:
import smtplib

In [4]:
email_id = "xyz@gmail.com"
pwd = "dummy-pwd"
to = "sender email"
subject = "test subject"

# Creating the server with host and port
server = smtplib.SMTP('smtp.gmail.com',587)
server.starttls() # Secure the communication

# For SSL, comment the above 2 lines and uncomment the below line
# server = smtplib.SMTP_SSL('smtp.gmail.com',465)
server.login(email_id,pwd)
server.sendmail(email_id,to,message)
server.quit()

SMTPAuthenticationError: (535, b'5.7.8 Username and Password not accepted. Learn more at\n5.7.8  https://support.google.com/mail/?p=BadCredentials z126sm15469026pfc.94 - gsmtp')

<pre>
By default, you will get the error
<b>SMTPAuthenticationError: (535, b'5.7.8 Username and Password not accepted. Learn more at\n5.7.8  https://support.google.com/mail/?p=BadCredentials z126sm15469026pfc.94 - gsmtp')</b>

Google: less secure apps gmail
switch on this. We have to make the gmail less secure in order to send mail through smtp.
We have to turn off the 2 factor authentication also
</pre>

In [ ]:
from_email_id = "xyz@gmail.com"
pwd = "dummy-pwd"
to = "sender email"
subject = "test subject"
message = "Hello from Jupyter"

# Creating the server with host and port
server = smtplib.SMTP_SSL('smtp.gmail.com',465)
server.login(email_id,pwd)
# \r - carriage return - previous data gets appended
body = "\r\n".join(['To:{}'.format(to), 'From:{}'.format(from_email_id),
                    'Subject:{}'.format(subject), 'Message: {}'.format(message)])
try:
    server.sendmail(from_email_id,to,body)
    print("Mail sent")
except Exception as e:
    print(str(e))
server.quit()